src/Feldspar/Onnx/onnxToFeld.hs

```diff
@@ -99,49 +104,49 @@ buildValues TD.DOUBLE     t = D.foldMap (\ x -> B.doubleDec x      <> B.string8
 buildValues TD.UINT64     t = D.foldMap (\ x -> B.word64Dec x      <> B.string8 "\n") $ TP.uint64_data t
 buildValues TD.INT64      t = D.foldMap (\ x -> B.int64Dec x       <> B.string8 "\n") $ TP.int64_data t
-buildValues TD.STRING     t = D.foldMap (\ x -> B.lazyByteString x     <> B.string8 "\n") $ TP.string_data t
-buildValues td            _ = error $ "onnxToFeld.buildValues: unsupported element type " ++ showElemT td
+buildValues TD.STRING     t = D.foldMap (\ x -> B.lazyByteString x <> B.string8 "\n") $ TP.string_data t
+buildValues td            _ = error $ "onnxToFeld.buildValues: unsupported element type " ++ U.toString (showElemT td)
 
 -- | Construct a Feldspar program corresponding to the ONNX graph
-mkProgramFile :: G.GraphProto -> String
+mkProgramFile :: G.GraphProto -> L.ByteString
 mkProgramFile gr 
-  = unlines [ "module Main where"
+  = L.unlines [ "module Main where"
             , ""
             , "import Feldspar"
             , "import Feldspar.Compiler (program)"
             , "import Feldspar.Onnx.Operators"
             , ""
-            , "main = program " ++ name
+            , "main = program " <> name
             , ""
-            , name ++ " " ++ unwords params ++ " = " ++ tuplify (map (mangle . vipName) $ D.toList $ G.output gr)
+            , name <> " " <> L.unwords params <> " = " <> tuplify (map (mangle . vipName) $ D.toList $ G.output gr)
             , "  where "
-              ++ intercalate "\n        " ("-- Nodes " : concatMap mkNode nodes)
+              <> L.intercalate "\n        " ("-- Nodes " : concatMap mkNode nodes)
             ]
   where name = "model" -- strFromJ $ G.name gr
         wName = "weights"
-        initVs = map (toStr . fromJust . TP.name) $ D.toList $ G.initializer gr
+        initVs = map (strFromJ . TP.name) $ D.toList $ G.initializer gr
         initSet = S.fromList initVs
         params = wName : map mkParam inps
         mkParam p = mangle $ vipName p
         inps = filter (\ p -> vipName p `S.notMember` initSet) $ D.toList $ G.input gr
         nodes = D.toList $ G.node gr
 
 -- | Construct a Feldspar pattern binding for an ONNX graph node
-mkNode :: N.NodeProto -> [String]
-mkNode n = [outs ++ " = " ++ op ++ " " ++ attrs]
-         ++ [ "    " ++ mangle (toStr v) | v <- D.toList $ N.input n]
+mkNode :: N.NodeProto -> [L.ByteString]
+mkNode n = [outs <> " = " <> op <> " " <> attrs]
+         ++ [ "    " <> mangle (unUtf8 v) | v <- D.toList $ N.input n]
          ++ [""]
-  where outs = tuplify $ map (mangle . toStr) $ D.toList $ N.output n
-        op = "onnx" ++ toStr (fromJust $ N.op_type n) ++ "_" ++ show (length $ D.toList $ N.input n)
-        attrs = "[" ++ intercalate ", " (map showAttribute $ D.toList $ N.attribute n) ++ "]"
+  where outs = tuplify $ map (mangle . unUtf8) $ D.toList $ N.output n
+        op = "onnx" <> strFromJ (N.op_type n) <> "_" <> show (length $ D.toList $ N.input n)
```

Comment:
There is a `D.length` that lets us avoid converting things to a list that we just take the length of.

Reply:
Ok!